src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java

```diff
@@ -252,44 +252,56 @@ private void addShutdownListenerToChannel(Channel channel, Boolean withRecovery)
                         errorBuilder.append(channelCause);
                         String errorString = errorBuilder.toString();
                         LOGGER.warn(errorString);
+
+                        var pinIdToChannelHolder = getChannelHolderByChannel(channel);
+                        if (pinIdToChannelHolder == null) return;
+
                         if (withRecovery && recoverableErrors.stream().anyMatch(errorString::contains)) {
-                            int channelNumber =  channel.getChannelNumber();
-                            var pinIdToChannelHolder = getChannelHolderByChannelNumber(channelNumber);
-                            if ((pinIdToChannelHolder != null) && (pinIdToChannelHolder.getValue().state() != ChannelState.SUBSCRIBING)) {
-                                recoverSubscriptionsOfChannel(channelNumber);
-                            }
+                            var pinId = pinIdToChannelHolder.getKey();
+                            var holder = pinIdToChannelHolder.getValue();
+                            recoverSubscriptionsOfChannel(pinId, channel, holder);
                         }
                     }
                 }
             }
         });
     }
 
-    private @Nullable Map.Entry<PinId, ChannelHolder> getChannelHolderByChannelNumber(int channelNumber) {
+    private @Nullable Map.Entry<PinId, ChannelHolder> getChannelHolderByChannel(Channel channel) {
         return channelsByPin
                 .entrySet()
                 .stream()
-                .filter(entry -> Objects.nonNull(entry.getValue().channel) && channelNumber == entry.getValue().channel.getChannelNumber())
+                .filter(entry -> channel == entry.getValue().channel)
                 .findAny()
                 .orElse(null);
     }
 
-    private void recoverSubscriptionsOfChannel(int channelNumber) {
+    private void recoverSubscriptionsOfChannel(@NotNull final PinId pinId, Channel channel, @NotNull final ChannelHolder holder) {
         channelChecker.execute(() -> {
             try {
-                var pinIdToChannelHolder = getChannelHolderByChannelNumber(channelNumber);
-                if (pinIdToChannelHolder != null) {
-                    PinId pinId = pinIdToChannelHolder.getKey();
-                    ChannelHolder channelHolder = pinIdToChannelHolder.getValue();
+                holder.lock.lock();
```

Comment:
The lock should be taken outside the try-finally block. Otherwise, we might call unlock if the locking was not successful

Reply:
You are right, sorry